esdb/projection_test.go

```diff
@@ -0,0 +1,351 @@
+package esdb_test
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/EventStore/EventStore-Client-Go/v3/esdb"
+	"github.com/stretchr/testify/require"
+)
+
+func ProjectionTests(t *testing.T, emptyDBClient *esdb.Client) {
+	t.Run("ProjectionTests", func(t *testing.T) {
+		projClient := esdb.NewProjectionClientFromExistingClient(emptyDBClient)
+
+		t.Run("createProjection", createProjection(projClient))
+		t.Run("deleteProjection", deleteProjection(projClient))
+		t.Run("updateProjection", updateProjection(projClient))
+		t.Run("enableProjection", enableProjection(projClient))
+		t.Run("disableProjection", disableProjection(projClient))
+		t.Run("resetProjection", resetProjection(projClient))
+		t.Run("getStateProjection", getStateProjection(projClient))
+		t.Run("getResultProjection", getResultProjection(projClient))
+	})
+}
+
+func createProjection(client *esdb.ProjectionClient) TestCall {
+	return func(t *testing.T) {
+		script, err := os.ReadFile("../resources/test/projection.js")
+		require.NoError(t, err)
+		name := NAME_GENERATOR.Generate()
+
+		err = client.Create(context.Background(), name, string(script), esdb.CreateProjectionOptions{})
+		require.NoError(t, err)
+
+		waitUntilProjectionStatusIs(t, 5*time.Minute, client, name, "Running")
+	}
+}
+
+func deleteProjection(client *esdb.ProjectionClient) TestCall {
+	return func(t *testing.T) {
+		script, err := os.ReadFile("../resources/test/projection.js")
+		require.NoError(t, err)
+		name := NAME_GENERATOR.Generate()
+
+		err = client.Create(context.Background(), name, string(script), esdb.CreateProjectionOptions{})
+		require.NoError(t, err)
+
+		waitUntilProjectionStatusIs(t, 5*time.Minute, client, name, "Running")
+
+		err = client.Disable(context.Background(), name, esdb.GenericProjectionOptions{})
+		require.NoError(t, err)
+
+		waitUntilProjectionStatusIs(t, 5*time.Minute, client, name, "Stopped")
+
+		// This race-condition is still happening even if it's marked as fixed: https://github.com/EventStore/EventStore/issues/2938
```

Comment:
Do we need to revisit this/track on a new linear issue?

Reply:
What meant is while the issue is fixed, we still need that mitigation code in the tests because `previous-lts` (22.10) version doesn't have the fix and is still used in our client test suite. When 24.10 release will be out, `previous-lts` will be 23.10 then that mitigation code will no longer be needed.

Reply:
Ah gotcha gotcha